### PR TITLE
ID-653 Enhanced Bucket Logging

### PR DESCRIPTION
**Pull request by tlangs**
### Jira Ticket: https://broadworkbench.atlassian.net/browse/ID-653

Rawls: https://github.com/broadinstitute/rawls/pull/2392

## Summary of changes:
When creating (or cloning) a workspace, a user is now given the option to enable enhanced bucket logging _without_ an Auth Domain. 

### What
- If the user selects an Auth Domain, the checkbox is automatically checked and disabled. The UI code also has functionality to require the bucket logging, but that isn't exercised in this PR.

### Why
- Requiring an Auth Domain just to get bucket logging is terrible UX. This is better. 

### Testing strategy
- [x] Create a new workspace, and check the "Enhanced Bucket Logging" checkbox. The new workspace bucket should begin with `fc-secure`

https://github.com/DataBiosphere/terra-ui/assets/3210510/21a60328-bcdc-4ad7-abe1-2a6871067810

### CODE REVIEW

**Comment by tlangs** [comment]
ebm_test
Tested on deployment

File: src/components/NewWorkspaceModal.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: `aria-label` is an alternative to having an explicit label for an input item (and if you need it, it should be real words). You have a label, so you shouldn't need `aria-label` here.

File: src/libs/forms.js
Comment (cahrens) [inline src/libs/forms.js]: Did you need this?

File: src/components/NewWorkspaceModal.js
Comment (tlangs) [reply src/components/NewWorkspaceModal.js]: Interesting. I was getting complaints in the console without it.

File: src/libs/forms.js
Comment (tlangs) [reply src/libs/forms.js]: Nope. Leftover and forgot to remove.

File: src/components/NewWorkspaceModal.js
Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: Then you aren't linking the label and checkbox together properly.

File: src/components/NewWorkspaceModal.js
Comment (calypsomatic) [inline src/components/NewWorkspaceModal.js]: Can we get a real link in here?

Comment (tlangs) [reply src/components/NewWorkspaceModal.js]: Absolutely before it gets merged! I don't think we have a support article for Enhanced Bucket Logging yet. How would we get that done?

File: src/components/NewWorkspaceModal.js
Comment (tlangs) [reply src/components/NewWorkspaceModal.js]: I'm not sure what that means...

File: src/components/NewWorkspaceModal.js
Comment (calypsomatic) [reply src/components/NewWorkspaceModal.js]: I guess ask your product person/user support?  But even without a full article maybe we can put 2-3 sentences here to give a rough idea?

Comment (tlangs) [reply src/components/NewWorkspaceModal.js]: Wrote a blurb.

File: src/components/NewWorkspaceModal.js
Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: I can take a look. I think the console complaint may be our own attempt at enforcing a11y instead of an actually accessibility issue.

I am however seeing an issue with a missing aria-role on the checkbox. I can also see if I can figure out what needs to be done for this, as we don't generally have this problem with checkboxes. Perhaps you just need to set the default checked state to false instead of undefined.

image

**Comment by cahrens** [comment]
Note the alignment issue of the checkbox and the label. 

image

**Comment by cahrens** [comment]
@tlangs a functionality question-- if the user **doesn't** select Enhanced Bucket Logging, but does select an Auth domain, do they still get "Enhanced Bucket Logging"? FYI @admull.

**Comment by tlangs** [comment]
@cahrens Yes they do! That's controlled by Rawls on workspace creation. Enhanced Bucket Logging just lets the user get the logging without the need to manage an Auth Domain, but Auth Domain functionality remains unchanged.

**Comment by admull** [comment]
Functionally, this is fine. I don't want to block this from merging if there's urgency here, but I'd really like us to see this as an opportunity to advance the user experience here instead of adding more fields to a screen with difficult to understand context.

File: src/libs/forms.js
Comment (cahrens) [inline src/libs/forms.js]: Shouldn't need this.

File: src/components/NewWorkspaceModal.js
Comment (calypsomatic) [inline src/components/NewWorkspaceModal.js]: Do users know the import of this, or should we mention that it's required for some protected data sets?

File: src/components/NewWorkspaceModal.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: To fix the a11y issue about the checked state not being set when initially rendered, make this `!!requireEnhancedBucketLogging` (so it is false instead of undefined).

File: src/components/NewWorkspaceModal.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: To fix the a11y issue about a label, use this.

I'm somewhat conflicted about having the InfoBox popup within the label… pros and cons either way.

```
isGoogleBillingProject() &&
                h(Fragment, [
                  div({ style: { padding: '.25rem' } }),
                  h(LabeledCheckbox, {
                    checked: enhancedBucketLogging,
                    disabled: requireEnhancedBucketLogging || groups.length > 0,
                    onChange: () => setEnhancedBucketLogging(!enhancedBucketLogging),
                    children: [
                      h(FormLabel, { style: { paddingLeft: '0.5rem', display: 'inherit' } }, [
                        'Enhanced Bucket Logging',
                        h(InfoBox, { style: { marginLeft: '0.25rem', padding: '0' } }, [
                          '"Enhanced Bucket Logging" means ...',
                          h(
                            Link,
                            {
                              href: 'https://youtu.be/E4WlUXrJgy4',
                              ...Utils.newTabLinkProps,
                            },
                            ['Read more about Enhanced Bucket Logging']
                          ),
                        ]),
                      ]),
                    ],
                  }),
                ]),

File: src/components/NewWorkspaceModal.js
Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: I added comments with fixes for both issues.

**Comment by cahrens** [comment]
> @cahrens Yes they do! That's controlled by Rawls on workspace creation. Enhanced Bucket Logging just lets the user get the logging without the need to manage an Auth Domain, but Auth Domain functionality remains unchanged.

This is what I'm concerned about… the user may think that they can fully control whether they get "Enhanced Bucket Logging" by using the checkbox, independent of the auth domain. We should at least make sure that the documentation states that all workspaces with auth domains also have "Enhanced Bucket Logging" irrespective of whether the checkbox is selected.

(Ideally I would expect that selecting an auth domain checks "Enhanced Bucket Logging" and disables it so that the user cannot deselect it)

Oh wait, it actually does what I wanted! I just hadn't tried it. Never mind!

**Comment by cahrens** [comment]
Chris' PR now merged, so you can rebase and pick up the new unittest file for `NewWorkspaceModal`. You can then add some tests about when the checkbox appears and how its state is tied to the auth domain input.

File: src/components/NewWorkspaceModal.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: Note that you can  (and should) drop the `IdContainer` now.

**Comment by tlangs** [comment]
@cahrens I just spent an hour tweaking the checkbox and essentially came to the same conclusion 😅. There are no more a11y warnings when the checkbox is rendered, so I think I've gotten it down correctly.

**Comment by tlangs** [comment]
Removed the `IdContainer` and added tests. Thanks @cjllanwarne for laying the groundwork!

**Comment by admull** [comment]
@tlangs I'm going to be out tomorrow, but I think Kate and Beth had some suggestions to add some more text to make it easier for users to understand and align with how we're thinking about presenting policies. I'm hoping it won't be a large ask. Can you follow-up with them tomorrow?

**Review by Shakespeared — APPROVED** [review]

**Review by davidangb — APPROVED** [review]

File: src/components/NewWorkspaceModal.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: Since "Access to data will be logged by Terra" is additional information that is useful to understand when selecting "Workspace will have protected data", I would suggest that you link it to the checkbox itself for the benefit of screen readers. This can be accomplished in the following manner (it's fine for a control to have both a label and additional description(s), and the user will be notified that there is additional content they can read), where `aria-describedby` is used to do the linking.

```
isGoogleBillingProject() &&
                div({ style: { margin: '1rem 0.25rem 0.25rem 0' } }, [
                  h(IdContainer, [
                    (id) =>
                      h(Fragment, [
                        h(
                          LabeledCheckbox,
                          {
                            style: { margin: '0rem 0.25rem 0.25rem 0' },
                            checked: enhancedBucketLogging,
                            disabled: !!requireEnhancedBucketLogging || groups.length > 0,
                            onChange: () => setEnhancedBucketLogging(!enhancedBucketLogging),
                            'aria-describedby': id,
                          },
                          [label({ style: { ...Style.elements.sectionHeader } }, ['Workspace will have protected data'])]
                        ),
                        h(InfoBox, { style: { marginLeft: '0.25rem', verticalAlign: 'middle' } }, [
                          'If checked, Terra will log all data access requests to the workspace bucket. ' +
                            'This feature is automatically enabled when a workspace is created with Authorization Domains.',
                        ]),
                        p({ id, style: { marginTop: '.25rem' } }, ['Access to data will be logged by Terra']),
                      ]),
                  ]),
                ]),
 ```               
              
Theoretically the same is true for "Authorization domain" and "Additional group management controls", but the custom select component makes it more difficult because `aria-describedby` won't be passed though to the underlying controls… and to be honest, I don't think "Additional group management controls" provides any useful information. So I wouldn't bother with that one.

File: src/components/NewWorkspaceModal.test.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.test.js]: It would be good to add a test that this checkbox is not present if an Azure billing project is selected.

File: src/components/NewWorkspaceModal.test.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.test.js:237]: In this test, it would be good to assert that the checkbox is not disabled.

(Nit: I can also imagine filling out all the fields, selecting the checkbox, clicking on "Create Workspace", and spying on what is sent for the Ajax request to ensure that the protected data option is sent, but I don't see that any of the existing tests go to that level of testing, and `NewWorkspaceModal` may not be in a state that would make such a test straightforward.)

**Comment by cahrens** [comment]
@Shakespeared this generally looks good-- I just have a few suggestions.

File: src/components/NewWorkspaceModal.js
Comment (Shakespeared) [reply src/components/NewWorkspaceModal.js]: Good eye for accessibility! Updated from your example. 👍

File: src/components/NewWorkspaceModal.test.js
Comment (Shakespeared) [reply src/components/NewWorkspaceModal.test.js:237]: Added an assertion here as suggested. 

Yeah, while I agree it would be good to test that the option is actually being sent, I am not sure this is the moment to add that test. I would prefer that to be a test for the Ajax request function itself so I'd be reluctant to add it here.

File: src/components/NewWorkspaceModal.js
Comment (calypsomatic) [inline src/components/NewWorkspaceModal.js:66]: `requireEnhancedBucketLogging` is supposed to be that the new workspace MUST have enhanced logging no matter what, right?  testing out this branch locally, if I select an authorization domain, then remove it, the checkbox becomes unchecked.  at least for my purposes, i'd want it to always be checked whether or not they have an authorization domain.  what's the intended behavior?

Comment (calypsomatic) [reply src/components/NewWorkspaceModal.js:66]: My [upcoming PR](https://github.com/DataBiosphere/terra-ui/pull/4046) relies on this one; I got the behavior I want by adding a `requireEnhancedBucketLogging` [check](https://github.com/DataBiosphere/terra-ui/blob/a75a77735aebaaeb1102ff4fc9ddf773322671d0/src/components/NewWorkspaceModal.js#L394) to whether the checkbox is checked or not.  Is that correct behavior?  I have this branch merged into mine but don't know this one should merge into dev first.

File: src/components/NewWorkspaceModal.test.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.test.js]: Nit: this comment isn't relevant.

File: src/components/NewWorkspaceModal.test.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.test.js]: Nit: this comment isn't relevant.

File: src/components/NewWorkspaceModal.test.js
Comment (cahrens) [inline src/components/NewWorkspaceModal.test.js:361]: Nit: it would be nice to check what the text of the checkbox is (since you are using getByRole and thus could get a different checkbox).

**Review by cahrens — APPROVED** [review]

File: src/components/NewWorkspaceModal.js
Comment (Shakespeared) [reply src/components/NewWorkspaceModal.js:66]: @tlangs can you confirm the expected behavior here?

Comment (Shakespeared) [reply src/components/NewWorkspaceModal.js:66]: Am I understanding correctly that you're seeing this behavior on a workspace that starts with enhanced bucket logging enabled by default, then if you select an auth domain, then deselect an auth domain, the checkbox becomes unchecked?

Comment (calypsomatic) [reply src/components/NewWorkspaceModal.js:66]: Yep!  I pass in `requireEnhancedBucketLogging = true` to the `NewWorkspaceModal` and then I am able to remove it by selecting and deselecting an auth domain.

Comment (Shakespeared) [reply src/components/NewWorkspaceModal.js:66]: Hm that does seem like a bug to me, let me take a look!

Comment (Shakespeared) [reply src/components/NewWorkspaceModal.js:66]: @calypsomatic please double check but I think I fixed this! It looks like the change from deselecting an auth domain was updating the checkbox despite it being disabled. I added an additional check for `requireEnhancedBucketLogging` so it should respect that now and have the expected behavior

Comment (calypsomatic) [reply src/components/NewWorkspaceModal.js:66]: Looks great!  Thank you!

**Review by calypsomatic — APPROVED** [review]

**Pull request merged**
